src/Resource.ts

```diff
@@ -0,0 +1,47 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import { CachedValue } from "./CachedValue";
+import { templateKeys } from "./constants";
+import { IResource } from "./IResource";
+import * as Json from "./JSON";
+import * as language from "./Language";
+import { TemplateScope } from "./TemplateScope";
+import { getChildTemplateForResourceObject } from "./templateScopes";
+
+/**
+ * This class represents a member of the "resources" section in a deployment template
+ */
+export class Resource implements IResource {
+    private readonly _childTemplate: CachedValue<TemplateScope | undefined> = new CachedValue<TemplateScope | undefined>();
+    private readonly _nameValueCache: CachedValue<Json.StringValue | undefined> = new CachedValue<Json.StringValue | undefined>();
+
+    constructor(
+        private readonly parentScope: TemplateScope,
+        public readonly resourceObject: Json.ObjectValue
+    ) { }
+
+    /**
+     * Convenient way of seeing what this object represents in the debugger, shouldn't be used for production code
+     */
+    public get __debugDisplay(): string {
+        return this.nameValue?.unquotedValue ?? '(unnamed resource)';
+    }
+
+    public get childDeployment(): TemplateScope | undefined {
+        return this._childTemplate.getOrCacheValue(() => {
+            return getChildTemplateForResourceObject(this.parentScope, this.resourceObject);
```

Comment:
>getChildTemplateForResourceObject [](start = 19, length = 33)

Would it make sense for this to be a method of the Resource class rather than an external utility method? Or does it get called with values that are not fields of this object?

Reply:
It would, in fact I tried to do that, but that causes a circular dependence Resource -> TemplateScope -> Resource.